File: mobile/android/app/src/main/java/com/whereswaldo/android/ui/groups/GroupJoinScreen.kt
package com.whereswaldo.android.ui.groups

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.tooling.preview.Preview
import com.whereswaldo.android.ui.designsystem.WaldoTheme
import com.whereswaldo.android.ui.designsystem.components.WaldoButton
import com.whereswaldo.android.ui.designsystem.components.WaldoStatusChip
import com.whereswaldo.android.ui.designsystem.components.WaldoStatusTone
import com.whereswaldo.android.ui.designsystem.components.WaldoTextField
import com.whereswaldo.android.ui.designsystem.components.WaldoTopBar

/**
 * The A5 join-group screen (001-api-contract.md §12.6, specs/003-android-client.md §12.2): code
 * entry (manual, or prefilled from the `waldo://group-join?code=…` deep link) plus a per-group
 * display name — **always shown** (unlike [CreateGroupScreen]'s conditional field): 005 §1 makes
 * `displayName` the caller's per-group nickname regardless of whether they already have a
 * profile, so it's always meaningful here, merely REQUIRED only when [needsDisplayName] (no
 * profile yet, §1.5.3). The manual code field is deliberately **not** pre-sanitized as the user
 * types — [GroupJoinViewModel.join] runs [GroupJoinCodeSanitizer] on submit either way, so typed
 * and deep-link-prefilled codes go through the identical gate (specs/003 §12.2: "validated by the
 * same pure normalization logic before any network call").
 */
@Composable
fun GroupJoinRoute(
    viewModel: GroupJoinViewModel,
    modifier: Modifier = Modifier,
    prefillCode: String = "",
    onJoined: () -> Unit = {},
) {
    val state by viewModel.state.collectAsState()
    GroupJoinScreen(
        state = state,
        needsDisplayName = viewModel.needsDisplayName,
        prefillCode = prefillCode,
        onJoin = viewModel::join,
        onJoined = onJoined,
        modifier = modifier,
    )
}

@Composable
fun GroupJoinScreen(
    state: GroupJoinUiState,
    modifier: Modifier = Modifier,
    needsDisplayName: Boolean = false,
    prefillCode: String = "",
    onJoin: (code: String, displayName: String?) -> Unit = { _, _ -> },
    onJoined: () -> Unit = {},
) {
    var code by remember { mutableStateOf(prefillCode) }
    var displayName by remember { mutableStateOf("") }

    LaunchedEffect(state.joined) {
        if (state.joined != null) onJoined()
    }

    Column(modifier = modifier.fillMaxSize()) {
        WaldoTopBar(title = "Join a group")

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(WaldoTheme.spacing.md),
            verticalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.sm),
        ) {
            WaldoTextField(
                value = code,
                onValueChange = { code = it },
                label = "Group code",
                placeholder = "XXXX-XXXX",
            )

            WaldoTextField(
                value = displayName,
                onValueChange = { displayName = it },
                label = if (needsDisplayName) "Your display name" else "Your display name (optional)",
            )

            state.validationError?.let { WaldoStatusChip(label = it, tone = WaldoStatusTone.Danger) }
            state.joinError?.let { WaldoStatusChip(label = it, tone = WaldoStatusTone.Danger) }

            WaldoButton(
                text = if (state.isJoining) "Joining…" else "Join group",
                enabled = !state.isJoining,
                onClick = { onJoin(code, displayName.ifBlank { null }) },
            )
        }
    }
}

@Preview(name = "Group join — light", showBackground = true)
@Composable
private fun GroupJoinScreenLightPreview() {
    WaldoTheme(darkTheme = false) {
        GroupJoinScreen(state = GroupJoinUiState(), prefillCode = "7F3K9QRZ")
    }
}

@Preview(name = "Group join — dark (error)", showBackground = true)
@Composable
private fun GroupJoinScreenDarkPreview() {
    WaldoTheme(darkTheme = true) {
        GroupJoinScreen(state = GroupJoinUiState(joinError = "That group code isn't valid."), needsDisplayName = true)
    }
}

File: mobile/ios/WaldoKit/Sources/WaldoKit/Screens/Groups/GroupJoinScreen.swift
import SwiftUI

/// specs/004-ios-client.md §3.4/§3.5 (001 §12.6) — composes ONLY design-system components. Also
/// the target of the `waldo://group-join?code=…` deep link AND, since specs/007, the
/// `https://{joinLinkHost}/g#CODE` universal link: `AppCoordinator.handleDeepLink(_:)` routes here
/// with `prefillCode` already normalized (empty when a valid https link carried no usable
/// fragment, 007 §4).
public struct GroupJoinScreen: View {
    @Environment(\.theme) private var theme
    @ObservedObject private var viewModel: GroupJoinViewModel
    @State private var code: String
    @State private var displayName: String = ""
    private let onJoined: (GroupSummary) -> Void

    public init(viewModel: GroupJoinViewModel, prefillCode: String = "", onJoined: @escaping (GroupSummary) -> Void) {
        self.viewModel = viewModel
        self._code = State(initialValue: prefillCode)
        self.onJoined = onJoined
    }

    public var body: some View {
        VStack(spacing: theme.spacing.lg) {
            WaldoNavBar("Join a group")
            content
            Spacer()
        }
        .background(theme.colors.surfaceVariant)
    }

    @ViewBuilder
    private var content: some View {
        switch viewModel.state {
        case .idle, .error:
            form
        case .joining:
            LoadingStateView(message: "Joining…")
        case .joined(let group):
            joinedView(group)
        }
    }

    private var form: some View {
        VStack(spacing: theme.spacing.md) {
            if case .error(let message) = viewModel.state {
                ErrorStateView(message: message)
            }
            WaldoTextField("Group code", text: $code, placeholder: "XXXX-XXXX")
            WaldoTextField("Your name for this group", text: $displayName, placeholder: "Noor")
            WaldoButton("Join group") {
                Task { await viewModel.join(rawCode: code, displayName: displayName) }
            }
        }
        .padding(.horizontal, theme.spacing.xl)
    }

    private func joinedView(_ group: GroupSummary) -> some View {
        VStack(spacing: theme.spacing.md) {
            EmptyStateView(title: "You're in!", message: "You've joined \(group.name).")
            WaldoButton("View group") { onJoined(group) }
        }
        .padding(.horizontal, theme.spacing.xl)
    }
}
